File: server/src/routes/admin/analytics.ts
/**
 * Admin Analytics Routes
 *
 * System-wide stats, metrics, and time series data for dashboards.
 */

import { Router, Request, Response } from 'express'
import { db } from '../../db/index.js'
import { asyncHandler } from '../../middleware/errorHandler.js'

const router = Router()

/**
 * GET /api/admin/analytics/overview
 * Get high-level system metrics
 */
router.get('/overview', asyncHandler(async (_req: Request, res: Response) => {
  const now = Date.now()
  const oneDayAgo = now - 24 * 60 * 60 * 1000
  const sevenDaysAgo = now - 7 * 24 * 60 * 60 * 1000

  // User counts
  const userStats = db.prepare(`
    SELECT
      COUNT(*) as total,
      SUM(CASE WHEN created_at > ? THEN 1 ELSE 0 END) as new_24h,
      SUM(CASE WHEN created_at > ? THEN 1 ELSE 0 END) as new_7d,
      SUM(CASE WHEN banned_at IS NOT NULL THEN 1 ELSE 0 END) as banned
    FROM users
  `).get(oneDayAgo, sevenDaysAgo) as {
    total: number
    new_24h: number
    new_7d: number
    banned: number
  }

  // Active users (have synapse activity)
  const activeUsers = db.prepare(`
    SELECT
      COUNT(DISTINCT CASE WHEN joined_at > ? THEN user_id END) as active_24h,
      COUNT(DISTINCT CASE WHEN joined_at > ? THEN user_id END) as active_7d
    FROM synapse_explorers
  `).get(oneDayAgo, sevenDaysAgo) as {
    active_24h: number
    active_7d: number
  }

  // Space/synapse stats
  const spaceStats = db.prepare(`
    SELECT
      COUNT(*) as total,
      SUM(CASE WHEN state = 'discovered' THEN 1 ELSE 0 END) as discovered,
      SUM(CASE WHEN state = 'being_solved' THEN 1 ELSE 0 END) as in_progress
    FROM spaces
  `).get() as {
    total: number
    discovered: number
    in_progress: number
  }

  // Ship stats
  const shipStats = db.prepare(`
    SELECT
      COUNT(*) as total,
      SUM(CASE WHEN state = 'idle' THEN 1 ELSE 0 END) as idle,
      SUM(CASE WHEN state IN ('exploring', 'solving', 'searching', 'traveling') THEN 1 ELSE 0 END) as deployed
    FROM agents
  `).get() as {
    total: number
    idle: number
    deployed: number
  }

  res.json({
    users: {
      total: userStats.total,
      new24h: userStats.new_24h,
      new7d: userStats.new_7d,
      active24h: activeUsers.active_24h,
      active7d: activeUsers.active_7d,
      banned: userStats.banned,
    },
    spaces: {
      total: spaceStats.total,
      discovered: spaceStats.discovered,
      inProgress: spaceStats.in_progress,
      discoveryRate: spaceStats.total > 0 ? (spaceStats.discovered / spaceStats.total * 100).toFixed(2) : '0',
    },
    ships: {
      total: shipStats.total,
      idle: shipStats.idle,
      deployed: shipStats.deployed,
    },
  })
}))

/**
 * GET /api/admin/analytics/economy
 * Get token economy metrics
 */
router.get('/economy', asyncHandler(async (_req: Request, res: Response) => {
  const economy = db.prepare(`
    SELECT
      SUM(points) as total_points,
      SUM(agentic_balance) as total_agentic,
      SUM(total_agi_earned) as total_agi,
      SUM(total_teneo_earned) as total_teneo,
      SUM(usdc_spent) as total_usdc_spent,
      AVG(points) as avg_points,
      AVG(total_agi_earned) as avg_agi
    FROM users
  `).get() as {
    total_points: number
    total_agentic: number
    total_agi: number
    total_teneo: number
    total_usdc_spent: number
    avg_points: number
    avg_agi: number
  }

  // User level distribution
  const levelDistribution = db.prepare(`
    SELECT user_level, COUNT(*) as count
    FROM users
    GROUP BY user_level
    ORDER BY user_level
  `).all() as Array<{ user_level: number; count: number }>

  res.json({
    circulation: {
      totalPoints: economy.total_points || 0,
      totalAgentic: economy.total_agentic || 0,
      totalAgi: economy.total_agi || 0,
      totalTeneo: economy.total_teneo || 0,
      totalUsdcSpent: economy.total_usdc_spent || 0,
    },
    averages: {
      points: Math.round(economy.avg_points || 0),
      agi: Math.round(economy.avg_agi || 0),
    },
    levelDistribution: levelDistribution.map(l => ({
      level: l.user_level,
      count: l.count,
    })),
  })
}))

/**
 * GET /api/admin/analytics/synapse-types
 * Get synapse type distribution and completion stats
 */
router.get('/synapse-types', asyncHandler(async (_req: Request, res: Response) => {
  const synapseStats = db.prepare(`
    SELECT
      synapse_type,
      COUNT(*) as total,
      SUM(CASE WHEN state = 'discovered' THEN 1 ELSE 0 END) as discovered,
      SUM(CASE WHEN state = 'being_solved' THEN 1 ELSE 0 END) as in_progress,
      AVG(points_accumulated) as avg_progress,
      SUM(agi_reward) as total_agi_rewards
    FROM spaces
    GROUP BY synapse_type
    ORDER BY
      CASE synapse_type
        WHEN 'minor' THEN 1
        WHEN 'complex' THEN 2
        WHEN 'deep' THEN 3
        WHEN 'core' THEN 4
        WHEN 'rare' THEN 5
        WHEN 'legendary' THEN 6
        WHEN 'unique' THEN 7
      END
  `).all() as Array<{
    synapse_type: string
    total: number
    discovered: number
    in_progress: number
    avg_progress: number
    total_agi_rewards: number
  }>

  res.json({
    synapseTypes: synapseStats.map(s => ({
      type: s.synapse_type,
      total: s.total,
      discovered: s.discovered,
      inProgress: s.in_progress,
      completionRate: s.total > 0 ? (s.discovered / s.total * 100).toFixed(2) : '0',
      avgProgress: Math.round(s.avg_progress || 0),
      totalAgiRewards: s.total_agi_rewards || 0,
    })),
  })
}))

/**
 * GET /api/admin/analytics/discovery-rate
 * Get discovery rate over time (for charts)
 */
router.get('/discovery-rate', asyncHandler(async (req: Request, res: Response) => {
  const days = Math.min(90, Math.max(7, parseInt(req.query.days as string) || 30))
  const now = Date.now()
  const startTime = now - days * 24 * 60 * 60 * 1000

  // Group discoveries by day
  const discoveries = db.prepare(`
    SELECT
      DATE(discovered_at / 1000, 'unixepoch') as date,
      COUNT(*) as count
    FROM spaces
    WHERE state = 'discovered' AND discovered_at > ?
    GROUP BY DATE(discovered_at / 1000, 'unixepoch')
    ORDER BY date
  `).all(startTime) as Array<{ date: string; count: number }>

  res.json({
    period: `${days} days`,
    data: discoveries.map(d => ({
      date: d.date,
      discoveries: d.count,
    })),
  })
}))

/**
 * GET /api/admin/analytics/user-growth
 * Get user registration over time
 */
router.get('/user-growth', asyncHandler(async (req: Request, res: Response) => {
  const days = Math.min(90, Math.max(7, parseInt(req.query.days as string) || 30))
  const now = Date.now()
  const startTime = now - days * 24 * 60 * 60 * 1000

  const growth = db.prepare(`
    SELECT
      DATE(created_at / 1000, 'unixepoch') as date,
      COUNT(*) as new_users
    FROM users
    WHERE created_at > ?
    GROUP BY DATE(created_at / 1000, 'unixepoch')
    ORDER BY date
  `).all(startTime) as Array<{ date: string; new_users: number }>

  // Calculate cumulative total
  const totalBefore = db.prepare(`
    SELECT COUNT(*) as count FROM users WHERE created_at <= ?
  `).get(startTime) as { count: number }

  let cumulative = totalBefore.count
  const dataWithCumulative = growth.map(g => {
    cumulative += g.new_users
    return {
      date: g.date,
      newUsers: g.new_users,
      totalUsers: cumulative,
    }
  })

  res.json({
    period: `${days} days`,
    data: dataWithCumulative,
  })
}))

/**
 * GET /api/admin/analytics/top-users
 * Get top users by various metrics
 */
router.get('/top-users', asyncHandler(async (req: Request, res: Response) => {
  const metric = (req.query.metric as string) || 'agi'
  const limit = Math.min(50, Math.max(10, parseInt(req.query.limit as string) || 20))

  const columnMap: Record<string, string> = {
    agi: 'total_agi_earned',
    points: 'points',
    usdc: 'usdc_spent',
    agentic: 'agentic_balance',
  }

  const column = columnMap[metric] || 'total_agi_earned'

  const topUsers = db.prepare(`
    SELECT id, wallet, user_level, ${column} as value
    FROM users
    WHERE banned_at IS NULL
    ORDER BY ${column} DESC
    LIMIT ?
  `).all(limit) as Array<{
    id: string
    wallet: string
    user_level: number
    value: number
  }>

  res.json({
    metric,
    users: topUsers.map((u, i) => ({
      rank: i + 1,
      id: u.id,
      wallet: u.wallet,
      userLevel: u.user_level,
      value: u.value,
    })),
  })
}))

export default router

File: server-go/internal/handlers/admin/analytics.go
package admin

import (
	"fmt"
	"time"

	"teneo/server-go/internal/database"

	"github.com/gofiber/fiber/v2"
)

// GetOverview returns high-level system metrics
func GetOverview(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	userCounts, err := store.Queries.GetOverviewUserCounts(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get user counts"})
	}

	agentCounts, err := store.Queries.GetOverviewAgentCounts(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get agent counts"})
	}

	spaceCounts, err := store.Queries.GetOverviewSpaceCounts(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get space counts"})
	}

	deployed := agentCounts.Total - agentCounts.Idle
	discoveryRate := "0.0"
	if spaceCounts.Total > 0 {
		discoveryRate = fmt.Sprintf("%.1f", float64(spaceCounts.Discovered)/float64(spaceCounts.Total)*100)
	}

	return c.JSON(fiber.Map{
		"users": fiber.Map{
			"total":    userCounts.TotalUsers,
			"new24h":   int64(0),
			"new7d":    int64(0),
			"active24h": userCounts.ActiveUsers,
			"active7d":  userCounts.ActiveUsers,
			"banned":   userCounts.BannedUsers,
		},
		"ships": fiber.Map{
			"total":    agentCounts.Total,
			"idle":     agentCounts.Idle,
			"deployed": deployed,
		},
		"spaces": fiber.Map{
			"total":         spaceCounts.Total,
			"discovered":    spaceCounts.Discovered,
			"inProgress":    spaceCounts.BeingExplored,
			"discoveryRate": discoveryRate,
		},
		"timestamp": time.Now().UnixMilli(),
	})
}

// GetEconomy returns token economy metrics
func GetEconomy(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	totals, err := store.Queries.GetUserEconomyTotals(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get economy totals"})
	}

	// Level distribution
	levelRows, err := store.Queries.GetUserLevelDistribution(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get level distribution"})
	}

	levelDist := make([]fiber.Map, len(levelRows))
	for i, lc := range levelRows {
		levelDist[i] = fiber.Map{"level": lc.UserLevel, "count": lc.Count}
	}

	// Compute averages
	userCount, _ := store.Queries.GetUserCount(ctx)
	avgPoints := float64(0)
	avgAgi := float64(0)
	if userCount > 0 {
		avgPoints = totals.TotalPoints / float64(userCount)
		avgAgi = float64(totals.TotalAgi) / float64(userCount)
	}

	return c.JSON(fiber.Map{
		"circulation": fiber.Map{
			"totalPoints":    totals.TotalPoints,
			"totalAgentic":   totals.TotalAgentic,
			"totalAgi":       totals.TotalAgi,
			"totalTeneo":     totals.TotalTeneo,
			"totalUsdcSpent": totals.TotalUsdc,
		},
		"averages": fiber.Map{
			"points": avgPoints,
			"agi":    avgAgi,
		},
		"levelDistribution": levelDist,
		"timestamp":         time.Now().UnixMilli(),
	})
}

// GetSynapseTypes returns synapse type distribution
func GetSynapseTypes(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	stats, err := store.Queries.GetSynapseTypeDistribution(ctx)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get synapse types"})
	}

	types := make([]fiber.Map, len(stats))
	for i, s := range stats {
		completionRate := "0.0"
		if s.Total > 0 {
			completionRate = fmt.Sprintf("%.1f", float64(s.Discovered)/float64(s.Total)*100)
		}
		types[i] = fiber.Map{
			"type":            s.SynapseType,
			"total":           s.Total,
			"discovered":      s.Discovered,
			"inProgress":      s.BeingExplored,
			"completionRate":  completionRate,
			"avgProgress":     0,
			"totalAgiRewards": 0,
		}
	}

	return c.JSON(fiber.Map{
		"synapseTypes": types,
		"timestamp":    time.Now().UnixMilli(),
	})
}

// GetDiscoveryRate returns discovery rate over time
func GetDiscoveryRate(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	days := c.QueryInt("days", 7)
	now := time.Now()
	startTime := now.AddDate(0, 0, -days).UnixMilli()

	dayCounts, err := store.Queries.GetDiscoveryRateByDay(ctx, &startTime)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get discovery rate"})
	}

	data := make([]fiber.Map, len(dayCounts))
	for i, dc := range dayCounts {
		data[i] = fiber.Map{"day": dc.Day, "count": dc.Count}
	}

	return c.JSON(fiber.Map{
		"discoveryRate": data,
		"days":          days,
		"timestamp":     time.Now().UnixMilli(),
	})
}

// GetUserGrowth returns user registration trends
func GetUserGrowth(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	days := c.QueryInt("days", 30)
	now := time.Now()
	startTime := now.AddDate(0, 0, -days).UnixMilli()

	dayCounts, err := store.Queries.GetUserGrowthByDay(ctx, startTime)
	if err != nil {
		return c.Status(500).JSON(fiber.Map{"error": "Failed to get user growth"})
	}

	data := make([]fiber.Map, len(dayCounts))
	cumulative := int64(0)
	for i, dc := range dayCounts {
		cumulative += dc.Count
		data[i] = fiber.Map{"day": dc.Day, "new": dc.Count, "cumulative": cumulative}
	}

	return c.JSON(fiber.Map{
		"userGrowth": data,
		"days":       days,
		"timestamp":  time.Now().UnixMilli(),
	})
}

// GetTopUsers returns top users by metric
func GetTopUsers(c *fiber.Ctx) error {
	store := c.Locals("store").(*database.Store)
	ctx := c.Context()

	metric := c.Query("metric", "agi")
	limit := int32(c.QueryInt("limit", 10))

	type topUser struct {
		ID             string
		Wallet         string
		TotalAgiEarned int32
		Points         float64
		UsdcSpent      float64
		AgenticBalance int32
	}

	var users []topUser

	switch metric {
	case "points":
		rows, err := store.Queries.GetTopUsersByPoints(ctx, limit)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to get top users"})
		}
		for _, r := range rows {
			users = append(users, topUser{ID: r.ID, Wallet: r.Wallet, TotalAgiEarned: r.TotalAgiEarned, Points: r.Points, UsdcSpent: r.UsdcSpent, AgenticBalance: r.AgenticBalance})
		}
	case "usdc":
		rows, err := store.Queries.GetTopUsersByUSDC(ctx, limit)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to get top users"})
		}
		for _, r := range rows {
			users = append(users, topUser{ID: r.ID, Wallet: r.Wallet, TotalAgiEarned: r.TotalAgiEarned, Points: r.Points, UsdcSpent: r.UsdcSpent, AgenticBalance: r.AgenticBalance})
		}
	case "agentic":
		rows, err := store.Queries.GetTopUsersByAgentic(ctx, limit)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to get top users"})
		}
		for _, r := range rows {
			users = append(users, topUser{ID: r.ID, Wallet: r.Wallet, TotalAgiEarned: r.TotalAgiEarned, Points: r.Points, UsdcSpent: r.UsdcSpent, AgenticBalance: r.AgenticBalance})
		}
	default: // "agi"
		rows, err := store.Queries.GetTopUsersByAGI(ctx, limit)
		if err != nil {
			return c.Status(500).JSON(fiber.Map{"error": "Failed to get top users"})
		}
		for _, r := range rows {
			users = append(users, topUser{ID: r.ID, Wallet: r.Wallet, TotalAgiEarned: r.TotalAgiEarned, Points: r.Points, UsdcSpent: r.UsdcSpent, AgenticBalance: r.AgenticBalance})
		}
	}

	result := make([]fiber.Map, len(users))
	for i, user := range users {
		var value float64
		switch metric {
		case "agi":
			value = float64(user.TotalAgiEarned)
		case "points":
			value = user.Points
		case "usdc":
			value = user.UsdcSpent
		case "agentic":
			value = float64(user.AgenticBalance)
		default:
			value = float64(user.TotalAgiEarned)
		}

		result[i] = fiber.Map{
			"rank":   i + 1,
			"id":     user.ID,
			"wallet": user.Wallet,
			"value":  value,
		}
	}

	return c.JSON(fiber.Map{
		"topUsers":  result,
		"metric":    metric,
		"timestamp": time.Now().UnixMilli(),
	})
}
